File: relisoft-hr/server/controllers/workspaceController.js
import { ROLES, getRoleByValue, getRoleLabel, isHrRole, isManagerRole } from '../utils/roleEnums.js';
import Employee from '../models/Employee.js';
import User from '../models/User.js';
import Project from '../models/Project.js';
import Team from '../models/Team.js';
import EmployeeTeam from '../models/EmployeeTeam.js';
import OrganizationRole from '../models/OrganizationRole.js';
import LeaveType from '../models/LeaveType.js';
import LeaveBalance from '../models/LeaveBalance.js';
import HrPolicy from '../models/HrPolicy.js';
import bcrypt from 'bcryptjs';

export const getWorkspace = async (req, res) => {
  try {
    const employees = await Employee.find({ isActive: true })
      .populate('organizationRole')
      .populate({ path: 'primaryTeam', populate: { path: 'lead' } })
      .populate({
        path: 'userId',
        select: 'name email role roleValue employeeId department designation phone status',
      })
      .sort({ firstName: 1 });

    const employeeTeams = await EmployeeTeam.find()
      .populate({ path: 'team', populate: { path: 'project lead' } });

    const projects = await Project.find().sort({ name: 1 }).lean();

    const teams = await Team.find().populate('lead project').sort({ name: 1 }).lean();

    const leaveTypes = await LeaveType.find({ isActive: true }).sort({ name: 1 }).lean();

    const hrPolicy = await getOrCreateHrPolicy();

    const orgRoles = await OrganizationRole.find({ isActive: true }).lean();

    const employeeMap = {};
    for (const emp of employees) {
      const empTeams = employeeTeams.filter(et => et.employee?.toString() === emp._id.toString());
      emp._doc.teams = empTeams.map(et => ({
        id: et.team?._id,
        name: et.team?.name,
        projectName: et.team?.project?.name,
        leadName: et.team?.lead?.firstName + ' ' + (et.team?.lead?.lastName || ''),
      }));
      emp._doc.roleLabel = emp.roleValue ? getRoleLabel(emp.roleValue) : 'Employee';
      employeeMap[emp._id] = emp;
    }

    const employeeList = Object.values(employeeMap).map(emp => ({
      id: emp._id,
      employeeCode: emp.employeeCode,
      firstName: emp.firstName,
      lastName: emp.lastName,
      fullName: `${emp.firstName} ${emp.lastName || ''}`.trim(),
      email: emp.email,
      phone: emp.phone,
      department: emp.department,
      designation: emp.designation,
      jobRole: emp.jobRole,
      employmentType: emp.employmentType,
      workLocation: emp.workLocation,
      salaryStructureDetails: emp.salaryStructureDetails,
      dateOfJoining: emp.dateOfJoining,
      roleLabel: emp._doc.roleLabel,
      roleValue: emp.roleValue,
      primaryTeam: emp.primaryTeam ? {
        id: emp.primaryTeam._id,
        name: emp.primaryTeam.name,
        leadName: emp.primaryTeam.lead ? `${emp.primaryTeam.lead.firstName} ${emp.primaryTeam.lead.lastName || ''}`.trim() : null,
      } : null,
      teams: emp._doc.teams,
      organizationRole: emp.organizationRole,
      isActive: emp.isActive,
    }));

    res.json({
      employees: employeeList,
      projects: projects.map(p => ({
        id: p._id,
        name: p.name,
        teams: teams.filter(t => t.project?._id?.toString() === p._id.toString()).map(t => ({
          id: t._id,
          name: t.name,
          projectId: t.project?._id,
          leadId: t.lead?._id,
          leadName: t.lead ? `${t.lead.firstName} ${t.lead.lastName || ''}`.trim() : null,
        })),
      })),
      leaveTypes: leaveTypes.map(lt => ({
        id: lt._id,
        name: lt.name,
        isPaid: lt.isPaid,
        isCompOff: lt.isCompOff,
        isFloater: lt.isFloater,
        requiresAdvanceNotice: lt.requiresAdvanceNotice,
        advanceNoticeDays: lt.advanceNoticeDays,
        isAccrued: lt.isAccrued,
        accrualPerMonth: lt.accrualPerMonth,
      })),
      roles: Object.entries(ROLES).map(([key, val]) => ({
        id: val.value,
        name: key,
        label: val.label,
        baseRole: val.baseRole,
      })),
      orgRoles: orgRoles.map(r => ({
        id: r._id,
        name: r.name,
        label: r.label,
        baseRole: r.baseRole,
        roleValue: r.roleValue,
      })),
      hrPolicy: { allowHalfDayLeave: hrPolicy.allowHalfDayLeave },
    });
  } catch (error) {
    console.error('getWorkspace error:', error);
    res.status(500).json({ message: error.message });
  }
};

export const createEmployee = async (req, res) => {
  try {
    const {
      employeeCode, firstName, lastName, email, department, designation,
      jobRole, employmentType, workLocation, dateOfJoining, roleValue,
      projectName, primaryTeam, additionalTeams, teamLeadEmpcode,
      salaryStructureDetails, username, temporaryPassword,
    } = req.body;

    const errors = [];
    if (!employeeCode) errors.push('EmployeeCode is required');
    if (!firstName) errors.push('FirstName is required');
    if (!email) errors.push('Email is required');
    if (!department) errors.push('Department is required');
    if (!designation) errors.push('Designation is required');
    if (!jobRole) errors.push('JobRole is required');
    if (!employmentType) errors.push('EmploymentType is required');
    if (!workLocation) errors.push('Location is required');
    if (!roleValue) errors.push('Role is required');
    if (errors.length > 0) return res.status(400).json({ message: errors.join('; ') });

    const existing = await Employee.findOne({ employeeCode });
    if (existing) return res.status(400).json({ message: `Employee code '${employeeCode}' already exists` });

    const emailExists = await Employee.findOne({ email });
    if (emailExists) return res.status(400).json({ message: `Email '${email}' already exists` });

    let project = null;
    if (projectName) {
      project = await Project.findOneAndUpdate(
        { name: projectName.trim() },
        { name: projectName.trim() },
        { upsert: true, new: true }
      );
    }

    const role = getRoleByValue(roleValue);
    const emp = await Employee.create({
      employeeCode, firstName, lastName, email, department, designation,
      jobRole, employmentType, workLocation, dateOfJoining: dateOfJoining || new Date(),
      roleValue, salaryStructureDetails, isActive: true,
    });

    const teamNames = [primaryTeam].concat((additionalTeams || '').split(',').map(t => t.trim()).filter(Boolean))
      .filter((v, i, a) => a.indexOf(v) === i);

    let primaryTeamId = null;
    for (const tName of teamNames) {
      if (!tName) continue;
      let leadEmp = null;
      if (teamLeadEmpcode) leadEmp = await Employee.findOne({ employeeCode: teamLeadEmpcode });
      const team = await Team.findOneAndUpdate(
        { name: tName, project: project?._id },
        { name: tName, project: project?._id, lead: leadEmp?._id || emp._id },
        { upsert: true, new: true }
      );
      await EmployeeTeam.findOneAndUpdate(
        { employee: emp._id, team: team._id },
        { employee: emp._id, team: team._id },
        { upsert: true, new: true }
      );
      if (tName === primaryTeam) primaryTeamId = team._id;
    }

    if (primaryTeamId) {
      emp.primaryTeam = primaryTeamId;
      await emp.save();
    }

    const userLogin = username || employeeCode.toLowerCase();
    const tempPwd = temporaryPassword || 'demo123';
    const salt = await bcrypt.genSalt(10);
    const hashedPwd = await bcrypt.hash(tempPwd, salt);

    await User.findOneAndUpdate(
      { email },
      {
        name: `${firstName} ${lastName || ''}`.trim(),
        email,
        password: hashedPwd,
        role: role?.baseRole || 'employee',
        roleValue,
        employeeId: employeeCode,
        department,
        designation,
        username: userLogin,
        status: 'active',
      },
      { upsert: true }
    );

    const defaultBalances = {
      'Sick/Casual Leave': 7,
      'Planned Leave': 1,
    };
    const leaveTypes = await LeaveType.find({ name: { $in: Object.keys(defaultBalances) } });
    for (const lt of leaveTypes) {
      await LeaveBalance.findOneAndUpdate(
        { employee: emp._id, leaveType: lt._id },
        { employee: emp._id, leaveType: lt._id, allocatedLeaves: defaultBalances[lt.name] || 0, usedLeaves: 0 },
        { upsert: true }
      );
    }

    res.status(201).json({ success: true, message: `Employee ${firstName} ${lastName || ''} created`, employeeId: emp._id });
  } catch (error) {
    console.error('createEmployee error:', error);
    res.status(500).json({ message: error.message });
  }
};

export const updateEmployee = async (req, res) => {
  try {
    const { id } = req.params;
    const updates = req.body;

    const emp = await Employee.findByIdAndUpdate(id, updates, { new: true });
    if (!emp) return res.status(404).json({ message: 'Employee not found' });

    if (updates.email) {
      await User.findOneAndUpdate({ employeeId: emp.employeeCode }, { email: updates.email });
    }

    res.json({ success: true, employee: emp });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const createProject = async (req, res) => {
  try {
    const { name } = req.body;
    if (!name) return res.status(400).json({ message: 'Project name is required' });

    const existing = await Project.findOne({ name: name.trim() });
    if (existing) return res.status(400).json({ message: 'Project already exists' });

    const project = await Project.create({ name: name.trim() });
    res.status(201).json({ success: true, id: project._id, name: project.name });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const updateProject = async (req, res) => {
  try {
    const { id } = req.params;
    const { name } = req.body;
    if (!name) return res.status(400).json({ message: 'Project name is required' });

    const project = await Project.findByIdAndUpdate(id, { name: name.trim() }, { new: true });
    if (!project) return res.status(404).json({ message: 'Project not found' });

    res.json({ success: true, project });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const createTeam = async (req, res) => {
  try {
    const { name, projectId, leadEmployeeCode } = req.body;
    if (!name) return res.status(400).json({ message: 'Team name is required' });
    if (!projectId) return res.status(400).json({ message: 'Project is required' });

    const project = await Project.findById(projectId);
    if (!project) return res.status(404).json({ message: 'Project not found' });

    let lead = null;
    if (leadEmployeeCode) lead = await Employee.findOne({ employeeCode: leadEmployeeCode });

    const team = await Team.create({ name: name.trim(), project: projectId, lead: lead?._id || null });
    res.status(201).json({ success: true, id: team._id, name: team.name });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const updateTeam = async (req, res) => {
  try {
    const { id } = req.params;
    const { name, projectId, leadEmployeeCode } = req.body;

    const updates = {};
    if (name) updates.name = name.trim();
    if (projectId) updates.project = projectId;
    if (leadEmployeeCode) {
      const lead = await Employee.findOne({ employeeCode: leadEmployeeCode });
      if (lead) updates.lead = lead._id;
    }

    const team = await Team.findByIdAndUpdate(id, updates, { new: true }).populate('lead project');
    if (!team) return res.status(404).json({ message: 'Team not found' });

    res.json({ success: true, team });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const getProjects = async (req, res) => {
  try {
    const projects = await Project.find().sort({ name: 1 }).lean();
    const teams = await Team.find().populate('lead project').sort({ name: 1 }).lean();

    res.json(projects.map(p => ({
      id: p._id,
      name: p.name,
      teams: teams.filter(t => t.project?._id?.toString() === p._id.toString()).map(t => ({
        id: t._id,
        name: t.name,
        leadId: t.lead?._id,
        leadName: t.lead ? `${t.lead.firstName} ${t.lead.lastName || ''}`.trim() : null,
      })),
    })));
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const getHrPolicy = async (req, res) => {
  try {
    const policy = await getOrCreateHrPolicy();
    res.json({ allowHalfDayLeave: policy.allowHalfDayLeave });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export const updateHrPolicy = async (req, res) => {
  try {
    const { allowHalfDayLeave } = req.body;
    const policy = await HrPolicy.findOneAndUpdate(
      {},
      { allowHalfDayLeave: !!allowHalfDayLeave },
      { upsert: true, new: true }
    );
    res.json({ success: true, allowHalfDayLeave: policy.allowHalfDayLeave });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

async function getOrCreateHrPolicy() {
  let policy = await HrPolicy.findOne();
  if (!policy) {
    policy = await HrPolicy.create({ allowHalfDayLeave: false });
  }
  return policy;
}

export { ROLES, getRoleByValue, getRoleLabel };

File: relisoft-hr/server/Controllers/WorkspaceController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using RelisoftHR.Data;
using RelisoftHR.DTOs;
using RelisoftHR.Models;
using RelisoftHR.Services;

namespace RelisoftHR.Controllers;

[ApiController]
[Route("api/workspace")]
public class WorkspaceController : ControllerBase
{
    private readonly AppDbContext _db;
    private readonly ILeaveBalanceService _leaveBalanceService;
    private readonly JoinerAnnouncementService _joinerService;

    public WorkspaceController(AppDbContext db, ILeaveBalanceService leaveBalanceService, JoinerAnnouncementService joinerService)
    {
        _db = db;
        _leaveBalanceService = leaveBalanceService;
        _joinerService = joinerService;
    }

    private int GetUserId()
    {
        var claim = User.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value;
        return claim != null && int.TryParse(claim, out var id) ? id : 0;
    }

    [HttpGet]
    public async Task<ActionResult<WorkspaceResponse>> GetWorkspace()
    {
        var employees = await _db.Employees
            .Include(e => e.Role)
            .Include(e => e.SalaryStructure)
            .Include(e => e.PrimaryTeam).ThenInclude(t => t!.Project)
            .Include(e => e.EmployeeTeams).ThenInclude(et => et.Team).ThenInclude(t => t!.Project)
            .Include(e => e.LeaveBalances).ThenInclude(lb => lb.LeaveType)
            .ToListAsync();

        var projects = await _db.Projects
            .Include(p => p.Teams).ThenInclude(t => t.Lead)
            .ToListAsync();

        var leaveTypes = await _db.LeaveTypes.Where(lt => lt.IsActive).OrderBy(lt => lt.SortOrder).ToListAsync();
        var roles = await _db.OrganizationRoles.ToListAsync();
        var hrPolicy = await _db.HrPolicies.FirstOrDefaultAsync() ?? new HrPolicy();
        var employeeDtos = new List<EmployeeDto>();
        var plannedLeaveTypeId = leaveTypes.SingleOrDefault(leaveType => leaveType.Name == "Planned Leave")?.Id;
        foreach (var employee in employees)
            employeeDtos.Add(await MapEmployeeAsync(employee, plannedLeaveTypeId));

        return Ok(new WorkspaceResponse(
            employeeDtos,
            projects.Select(p => MapProject(p)).ToList(),
            leaveTypes.Select(lt => new LeaveTypeDto(lt.Id, lt.Name, lt.CarryForwardPct, lt.IsCompOff, lt.IsFloaterHoliday, lt.MaxFloaterPerYear, lt.CompOffValidityDays)).ToList(),
            roles.Select(r => new RoleDto(r.Id, r.Name, r.Label, r.IsCustom, r.BaseRoleId)).ToList(),
            new HrPolicyDto(hrPolicy.AllowHalfDayLeave, hrPolicy.SandwichLeave)
        ));
    }

    [HttpPost("employees")]
    public async Task<ActionResult<CreateEmployeeResponse>> CreateEmployee(CreateEmployeeRequest req)
    {
        // NEW: resolve the selected manager's EmployeeCode from their Id
        string? managerCode = null;
        if (req.ManagerId.HasValue)
        {
            var manager = await _db.Employees.FindAsync(req.ManagerId.Value);
            managerCode = manager?.EmployeeCode;
        }

        var employee = new Employee
        {
            EmployeeCode = req.EmployeeCode,
            FullName = req.FullName,
            Email = req.Email,
            Department = req.Department,
            Designation = req.Designation,
            JobRole = req.JobRole,
            EmploymentType = req.EmploymentType,
            Location = req.Location,
            JoinDate = req.JoinDate,
            RoleId = req.Role,
            PrimaryTeamId = req.PrimaryTeamId,
            ManagerCode = managerCode,   // NEW
            UanNumber = string.IsNullOrWhiteSpace(req.UanNumber) ? null : req.UanNumber.Trim(),
            PanNumber = string.IsNullOrWhiteSpace(req.PanNumber) ? null : req.PanNumber.Trim(),
            IsUnpaidIntern = req.IsUnpaidIntern
        };

        _db.Employees.Add(employee);
        await _db.SaveChangesAsync();

        if (req.SalaryStructure != null)
        {
            var ss = new SalaryStructure
            {
                EmployeeId = employee.Id,
                FixedPay = req.SalaryStructure.FixedPay,
                VariablePay = req.SalaryStructure.VariablePay,
                PF = req.SalaryStructure.PF,
                Gratuity = req.SalaryStructure.Gratuity,
                Insurance = req.SalaryStructure.Insurance,
                OtherDeductions = req.SalaryStructure.OtherDeductions
            };
            _db.SalaryStructures.Add(ss);
            await _db.SaveChangesAsync();
        }

        foreach (var teamId in req.TeamIds)
        {
            _db.EmployeeTeams.Add(new EmployeeTeam { EmployeeId = employee.Id, TeamId = teamId });
        }

        var username = req.Email.Split('@')[0];
        var tempPassword = "demo123";
        _db.UserLogins.Add(new UserLogin
        {
            EmployeeId = employee.Id,
            Username = username,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(tempPassword)
        });

        await _db.SaveChangesAsync();

        await _joinerService.AnnounceJoinerAsync(employee, GetUserId());

        return Ok(new CreateEmployeeResponse(
            "Employee registered successfully.",
            username, tempPassword
        ));
    }

    [HttpPut("employees/{id}")]
    public async Task<ActionResult> UpdateEmployee(int id, UpdateEmployeeRequest req)
    {
        var employee = await _db.Employees
            .Include(e => e.EmployeeTeams)
            .Include(e => e.SalaryStructure)
            .FirstOrDefaultAsync(e => e.Id == id);
        if (employee == null) return NotFound(new { message = "Employee not found." });

        employee.EmployeeCode = req.EmployeeCode;
        employee.FullName = req.FullName;
        employee.Email = req.Email;
        employee.Department = req.Department;
        employee.Designation = req.Designation;
        employee.JobRole = req.JobRole;
        employee.EmploymentType = req.EmploymentType;
        employee.Location = req.Location;
        employee.JoinDate = req.JoinDate;
        employee.RoleId = req.Role;
        employee.PrimaryTeamId = req.PrimaryTeamId;
        employee.UanNumber = string.IsNullOrWhiteSpace(req.UanNumber) ? null : req.UanNumber.Trim();
        employee.PanNumber = string.IsNullOrWhiteSpace(req.PanNumber) ? null : req.PanNumber.Trim();
        employee.IsUnpaidIntern = req.IsUnpaidIntern;
        employee.UpdatedOn = DateTime.UtcNow;

        if (req.SalaryStructure != null)
        {
            var ss = employee.SalaryStructure;
            if (ss == null)
            {
                ss = new SalaryStructure { EmployeeId = id };
                _db.SalaryStructures.Add(ss);
            }
            ss.FixedPay = req.SalaryStructure.FixedPay;
            ss.VariablePay = req.SalaryStructure.VariablePay;
            ss.PF = req.SalaryStructure.PF;
            ss.Gratuity = req.SalaryStructure.Gratuity;
            ss.Insurance = req.SalaryStructure.Insurance;
            ss.OtherDeductions = req.SalaryStructure.OtherDeductions;
        }

        _db.EmployeeTeams.RemoveRange(employee.EmployeeTeams);
        foreach (var teamId in req.TeamIds)
        {
            _db.EmployeeTeams.Add(new EmployeeTeam { EmployeeId = id, TeamId = teamId });
        }

        await _db.SaveChangesAsync();
        return Ok(new { message = "Employee updated." });
    }

    [HttpPost("projects")]
    public async Task<ActionResult> CreateProject(CreateProjectRequest req)
    {
        _db.Projects.Add(new Project { Name = req.Name });
        await _db.SaveChangesAsync();
        return Ok(new { message = "Project created." });
    }

    [HttpPut("projects/{id}")]
    public async Task<ActionResult> UpdateProject(int id, UpdateProjectRequest req)
    {
        var project = await _db.Projects.FindAsync(id);
        if (project == null) return NotFound(new { message = "Project not found." });
        project.Name = req.Name;
        await _db.SaveChangesAsync();
        return Ok(new { message = "Project updated." });
    }

    [HttpPost("teams")]
    public async Task<ActionResult> CreateTeam(CreateTeamRequest req)
    {
        _db.Teams.Add(new Team
        {
            Name = req.Name,
            ProjectId = req.ProjectId,
            LeadId = req.LeadId
        });
        await _db.SaveChangesAsync();
        return Ok(new { message = "Team created." });
    }

    [HttpPut("teams/{id}")]
    public async Task<ActionResult> UpdateTeam(int id, UpdateTeamRequest req)
    {
        var team = await _db.Teams.FindAsync(id);
        if (team == null) return NotFound(new { message = "Team not found." });
        team.Name = req.Name;
        team.ProjectId = req.ProjectId;
        team.LeadId = req.LeadId;
        await _db.SaveChangesAsync();
        return Ok(new { message = "Team updated." });
    }

    [HttpPut("hr-policy")]
    public async Task<ActionResult> UpdateHrPolicy(UpdateHrPolicyRequest req)
    {
        var policy = await _db.HrPolicies.FirstOrDefaultAsync();
        if (policy == null)
        {
            _db.HrPolicies.Add(new HrPolicy { AllowHalfDayLeave = req.AllowHalfDayLeave, SandwichLeave = req.SandwichLeave });
        }
        else
        {
            policy.AllowHalfDayLeave = req.AllowHalfDayLeave;
            policy.SandwichLeave = req.SandwichLeave;
            policy.UpdatedOn = DateTime.UtcNow;
        }
        await _db.SaveChangesAsync();
        return Ok(new { message = "HR policy updated." });
    }

    [HttpGet("delegates/{managerId}")]
    public async Task<ActionResult> GetDelegates(int managerId)
    {
        var delegates = await _db.ApprovalDelegates
            .Include(d => d.Delegate)
            .Include(d => d.Project)
            .Where(d => d.ManagerId == managerId)
            .ToListAsync();

        return Ok(delegates.Select(d => new { d.Id, d.DelegateId, DelegateName = d.Delegate?.FullName, d.ProjectId, ProjectName = d.Project?.Name, IsGeneral = !d.ProjectId.HasValue }));
    }

    [HttpPost("delegates")]
    public async Task<ActionResult> AddDelegate(ApprovalDelegateRequest req, [FromQuery] int managerId)
    {
        var exists = await _db.ApprovalDelegates.AnyAsync(d => d.ManagerId == managerId && d.ProjectId == req.ProjectId && d.DelegateId == req.DelegateId);
        if (exists) return Conflict(new { message = "Delegate already exists." });

        _db.ApprovalDelegates.Add(new ApprovalDelegate
        {
            ManagerId = managerId,
            ProjectId = req.ProjectId,
            DelegateId = req.DelegateId
        });
        await _db.SaveChangesAsync();
        return Ok(new { message = "Delegate added." });
    }

    [HttpDelete("delegates/{id}")]
    public async Task<ActionResult> RemoveDelegate(int id)
    {
        var d = await _db.ApprovalDelegates.FindAsync(id);
        if (d == null) return NotFound();
        _db.ApprovalDelegates.Remove(d);
        await _db.SaveChangesAsync();
        return Ok(new { message = "Delegate removed." });
    }

    [HttpGet("leave-report")]
    public async Task<ActionResult> GetLeaveReport([FromQuery] int? year)
    {
        var y = year ?? DateTime.UtcNow.Year;
        var leaves = await _db.LeaveApplications
            .Include(l => l.Employee)
                 .ThenInclude(e => e!.LeaveBalances)
            .Include(l => l.LeaveType)
            .Where(l => l.AppliedOn.Year == y)
            .OrderBy(l => l.Employee!.FullName)
            .ThenBy(l => l.AppliedOn)
            .ToListAsync();

        var report = new List<object>();

        foreach (var leave in leaves)
        {
            decimal remainingLeaves = 0;

            var balance = await _leaveBalanceService.GetBalanceAsync(
                leave.EmployeeId,
                leave.LeaveTypeId
            );

            if (balance != null)
            {
                remainingLeaves = balance.RemainingLeaves;
            }

            report.Add(new
            {
                leave.Id,
                EmployeeName = leave.Employee?.FullName,
                leave.Employee?.EmployeeCode,
                LeaveType = leave.LeaveType?.Name,
                leave.FromDate,
                leave.ToDate,
                leave.TotalDays,
                leave.Status,
                ApprovedBy = leave.ApproverName,
                RemainingLeaves = remainingLeaves,

                ExtraLeavesTaken =
                    leave.Status == "Approved"
                        ? leave.LopDays
                        : 0m,

                leave.LossOfPay
            });
        }

        return Ok(report);
    }

    [HttpGet("floater-usage/{employeeId}")]
    public async Task<ActionResult> GetFloaterUsage(int employeeId, [FromQuery] int year)
    {
        var y = year;
        var used = await _db.LeaveApplications
            .CountAsync(l => l.EmployeeId == employeeId && l.LeaveType!.IsFloaterHoliday && l.FromDate.Year == y && l.Status == "Approved");
        var lt = await _db.LeaveTypes.FirstOrDefaultAsync(l => l.IsFloaterHoliday);
        return Ok(new FloaterHolidayUsageDto(used, lt?.MaxFloaterPerYear ?? 2));
    }

    private async Task<EmployeeDto> MapEmployeeAsync(Employee e, int? plannedLeaveTypeId)
    {
        var teams = e.EmployeeTeams.Select(et => new TeamDto(
            et.Team!.Id, et.Team.Name, et.Team.ProjectId,
            et.Team.Project?.Name ?? "", et.Team.LeadId,
            et.Team.Lead?.FullName ?? ""
        )).ToList();

        SalaryStructureDto? ss = null;
        if (e.SalaryStructure != null)
        {
            ss = new SalaryStructureDto(e.SalaryStructure.FixedPay, e.SalaryStructure.VariablePay,
                e.SalaryStructure.PF, e.SalaryStructure.Gratuity,
                e.SalaryStructure.Insurance, e.SalaryStructure.OtherDeductions);
        }

        var leaveBalances = e.LeaveBalances.Select(lb => new LeaveBalanceDto(
            lb.Id, lb.LeaveTypeId, lb.LeaveType?.Name ?? "",
            lb.AllocatedLeaves, lb.UsedLeaves, lb.RemainingLeaves,
            lb.CarryForwardDays, lb.FinancialYear
        )).ToList();

        if (plannedLeaveTypeId.HasValue)
        {
            var plannedBalance = await _leaveBalanceService.GetBalanceAsync(e.Id, plannedLeaveTypeId.Value);
            if (plannedBalance != null)
            {
                var existingIndex = leaveBalances.FindIndex(balance => balance.LeaveTypeId == plannedLeaveTypeId.Value);
                var dynamicBalance = new LeaveBalanceDto(plannedBalance.Id, plannedBalance.LeaveTypeId,
                    plannedBalance.LeaveTypeName, plannedBalance.AllocatedLeaves, plannedBalance.UsedLeaves,
                    plannedBalance.RemainingLeaves);
                if (existingIndex >= 0) leaveBalances[existingIndex] = dynamicBalance;
                else leaveBalances.Add(dynamicBalance);
            }
        }

        return new EmployeeDto(
            e.Id, e.EmployeeCode, e.FullName, e.Email, e.Department,
            e.Designation, e.JobRole, e.EmploymentType, e.Status, e.Location,
            ss, e.JoinDate,
            e.Role?.Name ?? "", e.RoleId, e.Role?.Label,
            e.PrimaryTeam != null
                ? new TeamDto(e.PrimaryTeam.Id, e.PrimaryTeam.Name,
                    e.PrimaryTeam.ProjectId, e.PrimaryTeam.Project?.Name ?? "",
                    e.PrimaryTeam.LeadId, e.PrimaryTeam.Lead?.FullName ?? "")
                : null,
            e.PrimaryTeamId,
            teams,
            leaveBalances,
            null,
            e.UanNumber, e.PanNumber, e.IsUnpaidIntern
        );
    }

    private static ProjectDto MapProject(Project p)
    {
        return new ProjectDto(p.Id, p.Name,
            p.Teams.Select(t => new TeamDto(
                t.Id, t.Name, p.Id, p.Name, t.LeadId, t.Lead?.FullName ?? ""
            )).ToList()
        );
    }
}
